kernel/data-pipeline/dialect/mysql/src/main/java/org/apache/shardingsphere/data/pipeline/mysql/ingest/client/netty/MySQLBinlogEventPacketDecoder.java

```diff
@@ -69,11 +78,44 @@ protected void decode(final ChannelHandlerContext ctx, final ByteBuf in, final L
                 in.resetReaderIndex();
                 break;
             }
-            Optional.ofNullable(decodeEvent(payload, binlogEventHeader)).ifPresent(out::add);
+            AbstractBinlogEvent binlogEvent = decodeEvent(payload, binlogEventHeader);
+            if (null == binlogEvent) {
+                skipChecksum(binlogEventHeader.getEventType(), in);
+                return;
+            }
+            if (decodeWithTX) {
+                processEventWithTX(binlogEvent, out);
+            } else {
+                processEventIgnoreTX(binlogEvent, out);
+            }
             skipChecksum(binlogEventHeader.getEventType(), in);
         }
     }
     
+    private void processEventWithTX(final AbstractBinlogEvent binlogEvent, final List<Object> out) {
+        if (binlogEvent instanceof QueryEvent) {
+            QueryEvent queryEvent = (QueryEvent) binlogEvent;
+            if (TX_BEGIN_SQL.equals(queryEvent.getSql())) {
+                records.clear();
+            }
```

Comment:
When begin SQL reach, `records` might not be empty, could we just clear it?

Reply:
yes, because begin SQL reach, means a new transaction's arrives, but i will improve it, use `records = new LinkedList()` instead `clear()`